### PR TITLE
Workaround in RPC.Eth.EthGetCode to allow [] as return value from server

### PR DESCRIPTION
**Pull request by webwarrior-ws**
Some servers, like `nodes.mewapi.io/rpc/eth` and `www.ethercluster.com/etc`,
return empty array instead of "0x" for non-contract addresses.
Example: `{"jsonrpc":"2.0","id":1,"result":[]}`.
Handle that case and return "0x".
In case of non-empty array, throw UnexpectedReplyException.

Upstream PR: #866

### CODE REVIEW

File: src/Nethereum.RPC/Eth/EthGetCode.cs [hunk 1]
```diff
@@ -49,12 +49,17 @@ public EthGetCode(IClient client) : base(client, ApiMethods.eth_getCode.ToString
...
-        public Task<string> SendRequestAsync(string address,
+        public async Task<string> SendRequestAsync(string address,
                                              object id = null,
                                              CancellationToken cancellationToken = default(CancellationToken))
         {
             if (address == null) throw new ArgumentNullException(nameof(address));
-            return base.SendRequestAsync(id, cancellationToken, address.EnsureHexPrefix(), DefaultBlock);
+            var res = await this.Client.SendRequestAsync<object>(this.BuildRequest(address.EnsureHexPrefix(), DefaultBlock, id));
+            // Some servers may return empty array for non-contract addresses instead of "0x", so check for array
+            if (res is Newtonsoft.Json.Linq.JArray)
```
Comment (knocte) [inline src/Nethereum.RPC/Eth/EthGetCode.cs:59]: @webwarrior-ws you also need to check if the array is empty

File: src/Nethereum.RPC/Eth/EthGetCode.cs [hunk 2]
```diff
@@ -49,12 +49,17 @@ public EthGetCode(IClient client) : base(client, ApiMethods.eth_getCode.ToString
             return base.SendRequestAsync(id, cancellationToken, address.EnsureHexPrefix(), block);
         }
 
-        public Task<string> SendRequestAsync(string address,
+        public async Task<string> SendRequestAsync(string address,
```
Comment (knocte) [inline src/Nethereum.RPC/Eth/EthGetCode.cs:52]: @webwarrior-ws is this change related to this PR?

Comment (webwarrior-ws) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:52]: yes

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 1 above)
Comment (webwarrior-ws) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:59]: What exception to throw when it is not?

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 2 above)
Comment (knocte) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:52]: if it was working and compiling before without the async, why do we need it now for this change?

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 1 above)
Comment (knocte) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:59]: I don't know, UnexpectedReplyException? and then put the json in the message

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 2 above)
Comment (webwarrior-ws) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:52]: Because we need to await for result before we can check its type and act accordingly. The type of method doesn't change.

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 1 above)
Comment (webwarrior-ws) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:59]: JSON is unavailable at this location, only result value

File: src/Nethereum.RPC/Eth/EthGetCode.cs [hunk 3]
```diff
@@ -49,12 +49,17 @@ public EthGetCode(IClient client) : base(client, ApiMethods.eth_getCode.ToString
...
                                              object id = null,
                                              CancellationToken cancellationToken = default(CancellationToken))
         {
             if (address == null) throw new ArgumentNullException(nameof(address));
-            return base.SendRequestAsync(id, cancellationToken, address.EnsureHexPrefix(), DefaultBlock);
+            var res = await this.Client.SendRequestAsync<object>(this.BuildRequest(address.EnsureHexPrefix(), DefaultBlock, id));
```
Comment (knocte) [inline src/Nethereum.RPC/Eth/EthGetCode.cs]: @webwarrior-ws this should call `base.`, not `this.`

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 2 above)
Comment (knocte) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:52]: ah fair enough, I overlooked that

File: src/Nethereum.RPC/Eth/EthGetCode.cs (hunk 1 above)
Comment (webwarrior-ws) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:59]: Do we have `UnexpectedReplyException` somewhere or are you suggesting to define new type?

Comment (knocte) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:59]: I'm suggesting a new type unless you can find an adequate one

Comment (knocte) [reply src/Nethereum.RPC/Eth/EthGetCode.cs:59]: > JSON is unavailable at this location, only result value

The result value is an object representation of the JSON result, I'm sure it can be converted back to the original string somehow.

**Comment by knocte** [comment]
> So you want something like Exceptions.UnexpectedReplyException("Unexpected response value" + res.ToString(), res) ?

Only if that works (so test it first with a non-empty JArray).

**Comment by knocte** [comment]
> "Unexpected response value" + res.ToString(), res)

BTW don't forget the characters ": " to separate res.ToString() from the first part of the message.

**Comment by knocte** [comment]
Ok, this looks better code-wise. Now, nits about the commit message:
```
RPC.Eth: fix in EthGetCode to allow [] as return value

Some servers return empty array instead of "0x" for non-contract addresses, so handle that case
```

1. As it's said in CONTRIBUTING.md, please respect max length of title line and body lines.
2. As it's implied in CONTRIBUTING.md, the body of the commit message is a paragraph, not a title, so it should end with a '.'.
3. I'd prefer the word 'workaround' instead of 'fix'.
4. Please squash all commits into 1, since there is no need to have them separate.
5. Please put an example of a full JSON string of a return value from the picky servers, in the body of the commit message.
6. You only mentioned the picky servers in PR description, but it would be valuable to have them in commit message too (in fact, since this PR will probably end up with only 1 commit, as suggested above, there's no reason for PR description to not be exactly the same as the commit message).

**Comment by knocte** [comment]
Ok great, this looks good now and I was about to hit the button "Merge pull request", but I realized I should be looking at upstream to see if the bug also exists there, and indeed it does. So let's be good citizens, can you propose same PR into the original repo, not just in our fork.

**Comment by knocte** [comment]
https://github.com/Nethereum/Nethereum/pull/866#issuecomment-1216168101
